File: catapult/python/07_01_aggregate_bonded_transaction.py
import datetime
from binascii import unhexlify
from json import dumps, loads

import httpx
import websocket
from helper import encode_message, generate_deadline, get_config
from symbolchain.CryptoTypes import Hash256, PrivateKey
from symbolchain.facade.SymbolFacade import SymbolFacade
from symbolchain.sc import Amount
from symbolchain.symbol.Network import Network


def aggregate_bonded_transaction():
    config = get_config()
    url = config["private1"]["url"]
    raw_privatekey_1 = config["private1"]["address1"]["privatekey"]
    raw_privatekey_2 = config["private1"]["address2"]["privatekey"]
    raw_privatekey_3 = config["private1"]["address3"]["privatekey"]
    min_fee_multiplier = config["private1"]["minfeemultiplier"]
    mosaic_id = config["private1"]["workaddress"]["mosaicId"]

    # mijinにアクセスして、Cookie, GenerationHash, epochAdjustmentを取得する
    r = httpx.get(url + "/network/properties")
    cookies = r.cookies
    cookies_list = []
    for k, v in cookies.items():
        cookie_str = f"{k}={v}"
        cookies_list.append(cookie_str)
    cookie = "; ".join(f"{c}" for c in cookies_list)

    network = r.json()
    currency_mosaic_id = network["chain"]["currencyMosaicId"]
    currency_mosaic_id = currency_mosaic_id.replace("'", "")
    currency_mosaic_id = currency_mosaic_id[2:]
    generation_hash = network["network"]["generationHashSeed"]
    epoch = network["network"]["epochAdjustment"]
    epoch = epoch.replace("s", "")

    facade = SymbolFacade("mijinnet")
    facade.network = Network(
        "mijinnet",
        0x60,
        datetime.datetime.fromtimestamp(int(epoch)),
        Hash256(generation_hash),
    )

    # 送信元のアカウントをセット
    account_1 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_1)))
    account_2 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_2)))
    account_3 = facade.KeyPair(PrivateKey(unhexlify(raw_privatekey_3)))

    address_1 = facade.network.public_key_to_address(account_1.public_key)
    address_3 = facade.network.public_key_to_address(account_3.public_key)

    # トランザクションを作成するための内容を作る
    deadline = generate_deadline(epoch)
    fee = Amount(min_fee_multiplier)
    amount = Amount(1)

    transfer_transaction1 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_1.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_3,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx1"),
        }
    )

    transfer_transaction2 = facade.transaction_factory.create_embedded(
        {
            "signer_public_key": account_2.public_key,
            "type": "transfer_transaction_v1",
            "recipient_address": address_3,
            "mosaics": [
                {
                    "mosaic_id": int(mosaic_id, 16),
                    "amount": amount,
                }
            ],
            "message": encode_message("TEST MIJIN tx2"),
        }
    )

    # トランザクションを纏める
    embedded_transactions = [
        transfer_transaction1,
        transfer_transaction2,
    ]
    transactions_hash = facade.hash_embedded_transactions(embedded_transactions)

    bonded_transaction = facade.transaction_factory.create(
        {
            "type": "aggregate_bonded_transaction_v2",
            "signer_public_key": account_1.public_key,
            "deadline": deadline,
            "transactions_hash": transactions_hash,
            "transactions": embedded_transactions,
            "fee": fee,
        }
    )
    bonded_transaction.fee = Amount(0)
    bonded_signature = facade.sign_transaction(account_1, bonded_transaction)
    bonded_json_payload = facade.transaction_factory.attach_signature(
        bonded_transaction, bonded_signature
    )
    bonded_data = loads(bonded_json_payload)
    bonded_transaction_hash = facade.hash_transaction(bonded_transaction)
    signed_bonded_transaction = {
        "SignedTransaction": {
            "payload": bonded_data["payload"],
            "hash": str(bonded_transaction_hash),
            "signerPublicKey": str(account_1.public_key),
            "type": bonded_transaction.type_.value,
            "networkType": facade.network.identifier,
        }
    }
    print("------------------- signedTransaction ------------------------")
    print(dumps(signed_bonded_transaction, indent=2))

    hashlock_transaction = facade.transaction_factory.create(
        {
            "signer_public_key": account_1.public_key,
            "deadline": deadline,
            "type": "hash_lock_transaction_v1",
            "mosaic": {
                "mosaic_id": int(currency_mosaic_id, 16),
                "amount": 0,
            },
            "duration": 1000,
            "hash": bonded_transaction_hash,
        }
    )

    # 作成したトランザクションに署名する
    signature = facade.sign_transaction(account_1, hashlock_transaction)
    json_payload = facade.transaction_factory.attach_signature(
        hashlock_transaction, signature
    )
    data = loads(json_payload)
    transaction_hash = facade.hash_transaction(hashlock_transaction)
    signed_hashlock_transaction = {
        "SignedTransaction": {
            "payload": data["payload"],
            "hash": str(transaction_hash),
            "signerPublicKey": str(account_1.public_key),
            "type": hashlock_transaction.type_.value,
            "networkType": facade.network.identifier,
        }
    }

    # トランザクションID
    print("------------------- hashLockTransaction ------------------------")
    print(dumps(signed_hashlock_transaction, indent=2))
    print("------------------- sendTransaction ------------------------")

    headers = {"Content-Type": "application/json"}
    # ハッシュロックを先に送る
    httpx.put(url + "/transactions", headers=headers, json=data, cookies=cookies)

    # websocketを作成する
    def on_message(ws, message):
        json_data = loads(message)
        subscribe_topic = f"confirmedAdded/{address_1}"
        if "uid" in json_data:
            uid = json_data["uid"]
            body = dumps({"uid": uid, "subscribe": subscribe_topic})
            ws.send(body)

        if json_data.get("topic"):
            topic = json_data["topic"]
            if topic == subscribe_topic:
                print("------------------- Success --------------------------------")
                print("Success: Transaction")
                print(dumps(json_data["data"], indent=2))
                # 承認されたらAggregateBondedを送る
                httpx.put(
                    url + "/transactions/partial",
                    headers=headers,
                    json=bonded_data,
                    cookies=cookies,
                )
                print(
                    "Next Command: python ./07_02_aggregate_bonded_transaction_cosign.py address2",
                    bonded_transaction_hash,
                    bonded_data["payload"],
                )
                print("------------------------------------------------------------")

    ws_url = url.replace("https", "wss")
    ws = websocket.WebSocketApp(
        ws_url + "/ws",
        on_message=on_message,
        cookie=cookie,
    )
    try:
        ws.run_forever()
    except KeyboardInterrupt:
        pass
    except Exception as e:
        print(e)
    finally:
        ws.close()


if __name__ == "__main__":
    aggregate_bonded_transaction()

File: catapult/javascript/07_01_aggregate_bonded_transaction.js
const {
  Account,
  TransactionHttp,
  TransactionService,
  TransferTransaction,
  AggregateTransaction,
  HashLockTransaction,
  UInt64,
  Deadline,
  PlainMessage,
  RepositoryFactoryHttp,
  NamespaceId,
  Mosaic,
} = require("@tech-bureau/symbol-sdk");
const { ChronoUnit } = require("@js-joda/core");
const config = require("config");
const { firstValueFrom } = require("rxjs");

const fetch = require("node-fetch");
const fetchCookie = require("fetch-cookie");
const toughCookie = require("tough-cookie");

const url = config.get("private1.url");
const rawprivatekey1 = config.get("private1.address1.privatekey");
const rawprivatekey2 = config.get("private1.address2.privatekey");
const rawprivatekey3 = config.get("private1.address3.privatekey");

const minfeemultiplier = config.get("private1.minfeemultiplier");

//送信するMosaicのNamespace
const namespace = config.get("private1.workaddress.namespace");

// 送信するMosaic数
const sendmosaicamount = 1;

(async () => {
  // 一回mijinにアクセスして、Cookieを取得する
  const cookieJar = new toughCookie.CookieJar();
  const fetchCookieJar = fetchCookie(fetch, cookieJar);
  await fetchCookieJar(url);
  const cookie = await cookieJar.getCookieString(url);

  // Websocketオプション Cookieを設定する
  const websocketOptions = {
    headers: { cookie: cookie },
  };

  const repo = new RepositoryFactoryHttp(url, {
    fetchApi: fetchCookieJar,
    websocketOptions: websocketOptions,
  });

  //Networktypeをcatapultから取得
  const networkType = await firstValueFrom(repo.getNetworkType());
  //GenerationHashをcatapultから取得
  const ghash = await firstValueFrom(repo.getGenerationHash());

  //手数料の取得
  const tfee = await firstValueFrom(repo.createNetworkRepository().getTransactionFees());

  console.log("network", networkType);
  console.log("GenerationHash", ghash);
  console.log("Fee", tfee);

  const epoch = await firstValueFrom(repo.getEpochAdjustment());
  const currency = await firstValueFrom(repo.getCurrencies());

  //アカウント1をセット
  const account1 = Account.createFromPrivateKey(rawprivatekey1, networkType);
  const account2 = Account.createFromPrivateKey(rawprivatekey2, networkType);
  const account3 = Account.createFromPrivateKey(rawprivatekey3, networkType);

  const namespaceId = new NamespaceId(namespace);
  const mosaic = new Mosaic(namespaceId, UInt64.fromUint(sendmosaicamount));

  //トランザクションを作成するための内容を作る
  const transferTransaction1 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account3.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx1"), //メッセージ
    networkType
  );

  //トランザクションを作成するための内容を作る
  const transferTransaction2 = TransferTransaction.create(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間)
    account3.address, //送信先アドレス
    [mosaic], //送信するモザイク
    PlainMessage.create("TEST MIJIN tx2"), //メッセージ
    networkType
  );

  // トランザクションを纏め、AggregateTransactionを作成する。
  // (送信元が複数アドレスの処理なのでBondedを使用)
  const aggregteTransaction = AggregateTransaction.createBonded(
    Deadline.create(epoch), //いつまでトランザクションをもたせるか(default ２時間),
    [
      transferTransaction1.toAggregate(account1.publicAccount),
      transferTransaction2.toAggregate(account2.publicAccount),
    ],
    networkType,
    []
  ).setMaxFeeForAggregate(minfeemultiplier, 1);

  console.log(aggregteTransaction.maxFee);

  //作成したトランザクションに署名する
  const signedTransaction = account1.sign(aggregteTransaction, ghash);

  //この中にトランザクションIDが入ってる
  console.log("------------------- signedTransaction ------------------------");
  console.log(signedTransaction);

  //トランザクションをロックするトランザクションを作る
  const hashLockTransaction = HashLockTransaction.create(
    Deadline.create(epoch, 1, ChronoUnit.HOURS), //1時間
    currency.currency.createRelative(tfee.medianFeeMultiplier > 0 ? 10 : 0), // 10cat.carrency必要
    UInt64.fromUint(1000), // 1000block内で
    signedTransaction,
    networkType
  ).setMaxFee(minfeemultiplier);

  //署名する
  const signedHashLockTransaction = account1.sign(hashLockTransaction, ghash);

  console.log("------------------- hashRockTransaction ------------------------");
  console.log(hashLockTransaction);

  console.log("------------------- sendTransaction ------------------------");

  // const repo = new RepositoryFactoryHttp(url);
  const listener = repo.createListener();
  const transactionService = new TransactionService(repo.createTransactionRepository(), repo.createReceiptRepository());

  // トランザクションを送信する
  try {
    // websocketをopenする
    await listener.open();
    // 一旦ハッシュロックトランザクションを先に送る
    const hashLockSend = await firstValueFrom(transactionService.announce(signedHashLockTransaction, listener));
    // その次にAggregateBondedを送る
    const tx = await firstValueFrom(transactionService.announceAggregateBonded(signedTransaction, listener));
    console.log("------------------- Success --------------------------------");
    console.log("Success: Transaction", tx);
    console.log("Next Command: node ./07_02_aggregate_bonded_transaction_cosign address2", tx.transactionInfo.hash);
    console.log("------------------------------------------------------------");
    listener.close();
    return tx;
  } catch (e) {
    listener.close();
    e.code ? console.log("Failed TransactionStatusError", e.code) : console.log("Error: ", e);
    throw new Error(e);
  }
})();
